### PR TITLE
Making sure all tensors in `torch.cat` sequence have the same dtype.

### LINKED ISSUE #35014: `torch.cat` should do type promotion
## 🐛 Bug

When concatenating tensors of different dtypes, the result is unexpected (and often incorrect). This could be fixed by using type promotion.

## To Reproduce

```python
>>> xt = torch.ones((2, 1, 3), dtype=torch.int8)
>>> yt = torch.ones((2, 1, 3), dtype=torch.int32)
>>> torch.cat([xt, yt])
tensor([[[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 0, 0]],
        [[0, 1, 0]]], dtype=torch.int8)
>>> torch.cat([yt, xt])
tensor([[[        1,         1,         1]],
        [[        1,         1,         1]],
        [[ 16843009,       257, 180739632]],
        [[        1,  10605347, 805308415]]], dtype=torch.int32)
```

## Expected behavior

Here are possible behaviors:

### Solution 1. Raise an error.

This makes sense with the current implementation where there is no actual type promotion.
```python
>>> xt = torch.ones((2, 1, 3), dtype=torch.int8)
>>> yt = torch.ones((2, 1, 3), dtype=torch.int32)
>>> torch.cat([xt, yt])
ValueError: Cannot concatenate tensors of different dtypes!
```

### Solution 2. Promote/Demote to the dtype of the first tensor
This solution will lose precision
Notice the output dtype in the examples below.

```python
>>> xt = torch.ones((2, 1, 3), dtype=torch.int8)
>>> yt = torch.ones((2, 1, 3), dtype=torch.int32)
>>> torch.cat([xt, yt])
tensor([[[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]]], dtype=torch.int8)  # Because the first element is int8
>>> torch.cat([yt, xt])
tensor([[[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]]], dtype=torch.int32)  # Because the first element is int32
```

### Solution 3. Promote to the dtype of the tensor with the highest hierarchy

```python
>>> xt = torch.ones((2, 1, 3), dtype=torch.int8)
>>> yt = torch.ones((2, 1, 3), dtype=torch.int32)
>>> torch.cat([xt, yt])
tensor([[[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]]], dtype=torch.int32)  # Because int32 > int8
>>> torch.cat([yt, xt])
tensor([[[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]],
        [[1, 1, 1]]], dtype=torch.int32)  # Because int32 > int8
```

## Environment

```console
$> python ./torch/utils/collect_env.py
Collecting environment information...
PyTorch version: 1.5.0a0+762be86
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.2
GCC version: Could not collect
CMake version: version 3.13.3

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] torch==1.5.0a0+762be86
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.5.0a0+66f2922
[conda] Could not collect
```

## Additional context

cc @ezyang @gchanan @zou3519 @SsnL

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35150 Making sure all tensors in `torch.cat` sequence have the same dtype.**

Fixes #35014

Differential Revision: [D20578589](https://our.internmc.facebook.com/intern/diff/D20578589)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -15003,6 +15003,33 @@ def test_cat_big(self, device):
         result = torch.cat(concat_list)
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    @onlyCPU
+    def test_cat_bad_dtypes(self, device):
```
Comment (gchanan) [inline test/test_torch.py]: why is this onlyCPU?

Comment (z-a-f) [reply test/test_torch.py]: Oh I was testing it locally w/o GPU -- forgot to change back

**Comment by z-a-f** [comment]
Local run results:

```console
$> python test/test_torch.py -- TestTorchDeviceTypeCUDA.test_cat_bad_dtypes_cuda TestTorchDeviceTypeCPU.test_cat_bad_dtypes_cpu
..
----------------------------------------------------------------------
Ran 2 tests in 0.735s

OK
```

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
@gchanan we need it in 1.5

**Pull request closed without merging**